// angular4-notify/angular4-notify.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('rxjs/Subject'), require('@angular/router'), require('rxjs/add/operator/filter')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', 'rxjs/Subject', '@angular/router', 'rxjs/add/operator/filter'], factory) :
	(factory((global['angular4-notify'] = {}),global._angular_core,global._angular_common,global.rxjs_Subject,global._angular_router));
}(this, (function (exports,_angular_core,_angular_common,rxjs_Subject,_angular_router) { 'use strict';

var Notification = (function () {
    function Notification() {
    }
    return Notification;
}());

var TYPE_INFO = 'info';
var TYPE_ERROR = 'error';
var TYPE_WARNING = 'warning';
var NotificationsService = (function () {
    function NotificationsService() {
        this.notifications = new rxjs_Subject.Subject();
    }
    /**
     * @param {?} notification
     * @return {?}
     */
    NotificationsService.prototype.pushNotification = function (notification) {
        this.notifications.next(notification);
    };
    /**
     * @param {?} text
     * @return {?}
     */
    NotificationsService.prototype.addInfo = function (text) {
        var /** @type {?} */ notification = new Notification();
        notification.message = text;
        notification.type = TYPE_INFO;
        this.pushNotification(notification);
    };
    /**
     * @param {?} text
     * @return {?}
     */
    NotificationsService.prototype.addError = function (text) {
        var /** @type {?} */ notification = new Notification();
        notification.message = text;
        notification.type = TYPE_ERROR;
        this.pushNotification(notification);
    };
    /**
     * @param {?} text
     * @return {?}
     */
    NotificationsService.prototype.addWarning = function (text) {
        var /** @type {?} */ notification = new Notification();
        notification.message = text;
        notification.type = TYPE_WARNING;
        this.pushNotification(notification);
    };
    return NotificationsService;
}());
NotificationsService.decorators = [
    { type: _angular_core.Injectable },
];
/**
 * @nocollapse
 */
NotificationsService.ctorParameters = function () { return []; };

var ContainerComponent = (function () {
    /**
     * @param {?} notificationsService
     * @param {?} router
     */
    function ContainerComponent(notificationsService, router) {
        this.notificationsService = notificationsService;
        this.router = router;
        this.notifications = [];
    }
    /**
     * @param {?} notification
     * @return {?}
     */
    ContainerComponent.prototype.render = function (notification) {
        this.notifications.push(notification);
    };
    /**
     * @param {?} notification
     * @return {?}
     */
    ContainerComponent.prototype.findSimilar = function (notification) {
        return this.notifications.find(function (existingNotification) {
            return existingNotification.message == notification.message
                && existingNotification.type == notification.type;
        });
    };
    /**
     * @return {?}
     */
    ContainerComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.sub = this.notificationsService.notifications
            .subscribe(function (n) {
            if (_this.findSimilar(n)) {
                return;
            }
            _this.render(n);
        });
        this.routerSub = this.router.events
            .filter(function (event) { return event instanceof _angular_router.NavigationStart; })
            .subscribe(function (e) {
            _this.notifications = [];
        });
    };
    /**
     * @return {?}
     */
    ContainerComponent.prototype.ngOnDestroy = function () {
        this.sub.unsubscribe();
        this.routerSub.unsubscribe();
    };
    /**
     * @param {?} n
     * @return {?}
     */
    ContainerComponent.prototype.onDismissNotify = function (n) {
        return this.notifications.splice(this.notifications.indexOf(n), 1);
    };
    return ContainerComponent;
}());
ContainerComponent.decorators = [
    { type: _angular_core.Component, args: [{
                selector: 'angular4-notify-notifications-container',
                template: "<div class=\"notifications-container\"> <angular4-notify-notification *ngFor=\"let notification of notifications\" [notification]=\"notification\" (dismissNotify)=\"onDismissNotify($event)\"></angular4-notify-notification> </div>",
                styles: [".notifications-container { width: 100%; } .notify { width: 100%; padding: 8px 16px; } .notify i { margin-right: 16px; vertical-align: middle; cursor: pointer; } .error { background-color: #ef4a4a; } .info { background-color: #60db8f; } .warning { background-color: #dbb460; }"]
            },] },
];
/**
 * @nocollapse
 */
ContainerComponent.ctorParameters = function () { return [
    { type: NotificationsService, },
    { type: _angular_router.Router, },
]; };

var NotificationComponent = (function () {
    function NotificationComponent() {
        this.dismissNotify = new _angular_core.EventEmitter();
    }
    /**
     * @return {?}
     */
    NotificationComponent.prototype.dismiss = function () {
        return this.dismissNotify.emit(this.notification);
    };
    return NotificationComponent;
}());
NotificationComponent.decorators = [
    { type: _angular_core.Component, args: [{
                selector: 'angular4-notify-notification',
                template: "<div class=\"notify {{notification.type}}\"> <a type=\"button\" class=\"close\"> <i class=\"material-icons\" (click)=\"dismiss()\">clear</i> </a> {{notification.message}} </div> ",
                styles: [".notifications-container { width: 100%; } .notify { width: 100%; padding: 8px 16px; } .notify i { margin-right: 16px; vertical-align: middle; cursor: pointer; } .error { background-color: #ef4a4a; } .info { background-color: #60db8f; } .warning { background-color: #dbb460; }"]
            },] },
];
/**
 * @nocollapse
 */
NotificationComponent.ctorParameters = function () { return []; };
NotificationComponent.propDecorators = {
    'notification': [{ type: _angular_core.Input },],
    'dismissNotify': [{ type: _angular_core.Output },],
};

var NotificationsModule = (function () {
    function NotificationsModule() {
    }
    /**
     * @return {?}
     */
    NotificationsModule.forRoot = function () {
        return {
            ngModule: NotificationsModule,
            providers: [
                NotificationsService
            ]
        };
    };
    return NotificationsModule;
}());
NotificationsModule.decorators = [
    { type: _angular_core.NgModule, args: [{
                imports: [
                    _angular_common.CommonModule
                ],
                declarations: [
                    ContainerComponent,
                    NotificationComponent
                ],
                exports: [
                    ContainerComponent,
                    NotificationComponent
                ]
            },] },
];
/**
 * @nocollapse
 */
NotificationsModule.ctorParameters = function () { return []; };

exports.NotificationsModule = NotificationsModule;
exports.ContainerComponent = ContainerComponent;
exports.NotificationComponent = NotificationComponent;
exports.NotificationsService = NotificationsService;

Object.defineProperty(exports, '__esModule', { value: true });

})));
